server/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala

```diff
@@ -0,0 +1,196 @@
+    *       `Kleisli[OptionT[F, *], ContextRequest[F, A], Response[F]]`.
+    *
+    * @param acquire Function of the [[Request]] to a `F[ContextRequest]` to
+    *        run each time a new [[Request]] is received.
+    *
+    * @param release Function to run on the termination of a Request/Response
+    *        interaction, in all cases. The first argument is the request
+    *        context, which is guaranteed to exist if acquire succeeds. The
+    *        second argument is the response context, which will only exist if
+    *        the generation of the `ContextResponse` is successful. The third
+    *        argument is the exit condition, either completed, canceled, or
+    *        error.
+    */
+  def bracketRequestResponseCaseRoutes_[F[_], A, B](
+      acquire: Request[F] => F[ContextRequest[F, A]]
+  )(release: (A, Option[B], ExitCase[Throwable]) => F[Unit])(implicit
+      F: Bracket[F, Throwable]): FullContextMiddleware[F, A, B] =
+    (bracketRoutes: Kleisli[OptionT[F, *], ContextRequest[F, A], ContextResponse[F, B]]) =>
+      Kleisli((request: Request[F]) =>
+        OptionT(
+          acquire(request).flatMap(contextRequest =>
+            bracketRoutes(contextRequest)
+              .foldF(release(contextRequest.context, None, ExitCase.Completed) *> F.pure(
+                None: Option[Response[F]]))(contextResponse =>
+                F.pure(Some(contextResponse.response.copy(body =
+                  contextResponse.response.body.onFinalizeCaseWeak(ec =>
+                    release(contextRequest.context, Some(contextResponse.context), ec))))))
+              .guaranteeCase {
+                case ExitCase.Completed =>
+                  F.unit
+                case otherwise =>
+                  release(contextRequest.context, None, otherwise)
+              })
+        ))
+
+  /** Bracket on the start of a request and the completion of processing the
+    * response ''body Stream''.
+    *
+    * @note $releaseWarning
+    *
+    * @note A careful reader might be wondering where the analogous `use`
+    *       parameter from `cats.effect.Bracket` has gone. The use of the
+    *       acquired resource is running the request, thus the `use` function
+    *       is actually just the normal context route function from http4s
+    *       `Kleisli[OptionT[F, *], ContextRequest[F, A], Response[F]]`.
+    *
+    * @param acquire Effect to run each time a request is received. The result
+    *        of it is put into a `ContextRequest` and passed to the underlying
+    *        routes.
+    *
+    * @param release Effect to run at the termination of each response body,
+    *        or on any error after `acquire` has run. Will always be called
+    *        exactly once if `acquire` is invoked, for each request/response.
+    */
+  def bracketRequestResponseCaseRoutes[F[_], A](
+      acquire: F[A]
+  )(release: (A, ExitCase[Throwable]) => F[Unit])(implicit
```

Comment:
There should probably be a version of this that works in terms of `Resource`. All the resources I would want to use with this middleware come as `Resource`s.

Reply:
@nigredo-tori Are you suggesting a variant where the `F` is `Resource`?

Reply:
There's no reason other than More Names this can't be added, right?

Reply:
I'm suggesting something like
```scala
def bracketRequestResponseResourceRoutes[F[_], A](
    resource: Resource[F, A]
)(implicit F: Bracket[F, Throwable]): ContextMiddleware[F, A]
```
`Resource` wraps a pair of  `acquire` and `release`. Libraries generally expose `Resource`s other than those, so it makes sense to provide support for this use case.

Reply:
@nigredo-tori 

Unless I'm misunderstanding you (which is entirely possible) the only possible implementation of that signature which makes sense is something like this,

```scala
  // I've not compile this, but I think it is correct

  def bracketRequestResponseResourceRoutes[F[_], A](
    resource: Resource[F, A]
  )(implicit F: Bracket[F, Throwable]): ContextMiddleware[F, A] = {
    (bracketRoutes: Kleisli[OptionT[F, *], ContextRequest[F], Response[F]]) =>
    Kleisli((request: Request[F]) => 
      resource.use(a =>
        bracketRoutes.run(ContextRequest(a, resquest))
      )
    )
  }
```

I'd be very happy to add that, but I want to make sure those semantics are what you want. The `release` function of the `Resource` will be run when the effect to generate the `Response[F]`, but _before_ `Response.body: Stream[F, Byte]` is run. Is that what you are asking for?

My original indent for the `BracketRequestResponse` middleware was to be able to attach a finalizer to termination of the `Response.body` xor the `F[Option[Response[F]]]` if the latter was empty or failed. Unfortunately there is no way to write that logic using `Resource` with the current encodings. This is because the `F[Option[Response[F]]]` doesn't currently compose with the `Response.body: Stream[F, Byte]` scope.

Reply:
I was thinking more along these lines:
```scala
(contextRoutes: ContextRoutes[F, A]) => {
  val contextRoutes0: ContextRoutes[F, (A, F[Unit])] =
    contextRoutes.contramap(_.map(_._1))
  bracketRequestResponseRoutes[F, (A, F[Unit])](resource.allocated)(_._2)(contextRoutes0)
}
```
I haven't compiled it, but the types should align. `Resource.allocated` is the key here - it transforms a `Resource` into something compatible with the `acquire`/`release` shape.

This would be sloppy with `ExitCase` handling, though, since `allocated` always releases with `ExitCase.Completed`. We might want to first add some kind of `Resource.allocatedCase` to cats-effect.

Reply:
Ah, that makes more sense. I didn't realize that `Resource` exposed a function like `allocated`.